specs/core/0_beacon-chain.md

```diff
@@ -1296,28 +1296,18 @@ def get_attesting_balance(state: BeaconState, attestations: List[PendingAttestat
     return get_total_balance(state, get_unslashed_attesting_indices(state, attestations))
 ```
 
-```python
-def get_crosslink_from_attestation_data(state: BeaconState, data: AttestationData) -> Crosslink:
-    return Crosslink(
-        epoch=min(data.target_epoch, state.current_crosslinks[data.shard].epoch + MAX_CROSSLINK_EPOCHS),
-        previous_crosslink_root=data.previous_crosslink_root,
-        crosslink_data_root=data.crosslink_data_root,
-    )
-```
-
 ```python
 def get_winning_crosslink_and_attesting_indices(state: BeaconState, epoch: Epoch, shard: Shard) -> Tuple[Crosslink, List[ValidatorIndex]]:
-    shard_attestations = [a for a in get_matching_source_attestations(state, epoch) if a.data.shard == shard]
-    shard_crosslinks = [get_crosslink_from_attestation_data(state, a.data) for a in shard_attestations]
+    shard_attestations = [a for a in get_matching_source_attestations(state, epoch) if a.data.crosslink.shard == shard]
     candidate_crosslinks = [
-        c for c in shard_crosslinks
+        c for c in [a.data.crosslink for a in shard_attestations]
```

Comment:
This double selection is strange and will certainly cause confusion to readers. I understand you did it so you can select crosslinks and then assign to a short var to a more manageable var to use in the boolean, but it's pretty non-obvious what is going on or why.

I suggest still doing the selection in a line above `crosslinks = [a.data.crosslink for a in shard_attestations]`
or make the if statement multiline like the following

```python
    candidate_crosslinks = [
        a.data.crosslink for a in shard_attestations
        if (
            hash_tree_root(state.current_crosslinks[shard]) in
            (a.data.crosslink.previous_crosslink_root, hash_tree_root(a.data.crosslink)
        )
```

Reply:
Addressed, and cleaned up `get_winning_crosslink_and_attesting_indices` :)